#### Problem Statement : 
Understand how the bank approves and refuses loan. Find out different patterns and represent the outcomes to help the bank reduce the credit risk and interest risk.

In [3]:
#importing warning to remove the warnins
import warnings
warnings.filterwarnings('ignore')

In [4]:
# importing the required libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

Working on Current Application

In [5]:
#reading the application_data 
current_application_data = pd.read_csv('application_data.csv')

In [6]:
current_application_data.head(10)

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   
5      100008       0         Cash loans           M            N   
6      100009       0         Cash loans           F            Y   
7      100010       0         Cash loans           M            Y   
8      100011       0         Cash loans           F            N   
9      100012       0    Revolving loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   
5               Y             0           99000.0    490495.5      27517.5   
6               Y             1          171000.0   1560726.0      41301.0   
7               Y             0          360000.0   1530000.0      42075.0   
8               Y             0          112500.0   1019610.0      33826.5   
9               Y             0          135000.0    405000.0      20250.0   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   
5  ...                 0                0                0                0   
6  ...                 0                0                0                0   
7  ...                 0                0                0                0   
8  ...                 0                0                0                0   
9  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   
5                        0.0                       0.0   
6                        0.0                       0.0   
7                        0.0                       0.0   
8                        0.0                       0.0   
9                        NaN                       NaN   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   
5                         0.0                        0.0   
6                         0.0                        1.0   
7                         0.0                        0.0   
8                         0.0                        0.0   
9                         NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0         

In [7]:
#Checking number of rows and columns in the current_application_data dataframe
current_application_data.shape

(307511, 122)

In [11]:
# check the column-wise info of the dataframe

current_application_data.info(verbose= True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Data columns (total 122 columns):
 #    Column                        Dtype  
---   ------                        -----  
 0    SK_ID_CURR                    int64  
 1    TARGET                        int64  
 2    NAME_CONTRACT_TYPE            object 
 3    CODE_GENDER                   object 
 4    FLAG_OWN_CAR                  object 
 5    FLAG_OWN_REALTY               object 
 6    CNT_CHILDREN                  int64  
 7    AMT_INCOME_TOTAL              float64
 8    AMT_CREDIT                    float64
 9    AMT_ANNUITY                   float64
 10   AMT_GOODS_PRICE               float64
 11   NAME_TYPE_SUITE               object 
 12   NAME_INCOME_TYPE              object 
 13   NAME_EDUCATION_TYPE           object 
 14   NAME_FAMILY_STATUS            object 
 15   NAME_HOUSING_TYPE             object 
 16   REGION_POPULATION_RELATIVE    float64
 17   DAYS_BIRTH                    int64  
 18   DA

In [12]:
# Check the summary for the numeric columns 

current_application_data.describe()

SK_ID_CURR         TARGET   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count  307511.000000  307511.000000  307511.000000      3.075110e+05   
mean   278180.518577       0.080729       0.417052      1.687979e+05   
std    102790.175348       0.272419       0.722121      2.371231e+05   
min    100002.000000       0.000000       0.000000      2.565000e+04   
25%    189145.500000       0.000000       0.000000      1.125000e+05   
50%    278202.000000       0.000000       0.000000      1.471500e+05   
75%    367142.500000       0.000000       1.000000      2.025000e+05   
max    456255.000000       1.000000      19.000000      1.170000e+08   

         AMT_CREDIT    AMT_ANNUITY  AMT_GOODS_PRICE  \
count  3.075110e+05  307499.000000     3.072330e+05   
mean   5.990260e+05   27108.573909     5.383962e+05   
std    4.024908e+05   14493.737315     3.694465e+05   
min    4.500000e+04    1615.500000     4.050000e+04   
25%    2.700000e+05   16524.000000     2.385000e+05   
50%    5.135310e+05   24903.000000     4.500000e+05   
75%    8.086500e+05   34596.000000     6.795000e+05   
max    4.050000e+06  258025.500000     4.050000e+06   

       REGION_POPULATION_RELATIVE     DAYS_BIRTH  DAYS_EMPLOYED  ...  \
count               307511.000000  307511.000000  307511.000000  ...   
mean                     0.020868  -16036.995067   63815.045904  ...   
std                      0.013831    4363.988632  141275.766519  ...   
min                      0.000290  -25229.000000  -17912.000000  ...   
25%                      0.010006  -19682.000000   -2760.000000  ...   
50%                      0.018850  -15750.000000   -1213.000000  ...   
75%                      0.028663  -12413.000000    -289.000000  ...   
max                      0.072508   -7489.000000  365243.000000  ...   

       FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  \
count     307511.000000     307511.000000     307511.000000     307511.000000   
mean           0.008130          0.000595          0.000507          0.000335   
std            0.089798          0.024387          0.022518          0.018299   
min            0.000000          0.000000          0.000000          0.000000   
25%            0.000000          0.000000          0.000000          0.000000   
50%            0.000000          0.000000          0.000000          0.000000   
75%            0.000000          0.000000          0.000000          0.000000   
max            1.000000          1.000000          1.000000          1.000000   

       AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  \
count               265992.000000              265992.000000   
mean                     0.006402                   0.007000   
std                      0.083849                   0.110757   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      4.000000                   9.000000   

       AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
count               265992.000000              265992.000000   
mean                     0.034362                   0.267395   
std                      0.204685                   0.916002   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      8.000000                  27.000000   

       AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
count              265992.000000               265992.000000  
mean                    0.265474                    1.899974  
std                     0.794056                    1.869295  
min                     0.000000                    0.000000  
25%              

In [13]:
# Finding columns containing null values

null_columns = current_application_data.columns[current_application_data.isnull().any()]  
emptycol=current_application_data.isnull().sum()
emptycol=emptycol[emptycol.values>(0.3*len(emptycol))]

In [14]:
#Calculating total columns having null values using len()

len(null_columns)

67

In [15]:
#Printing name of columns having null values
null_columns

Index(['AMT_ANNUITY', 'AMT_GOODS_PRICE', 'NAME_TYPE_SUITE', 'OWN_CAR_AGE',
       'OCCUPATION_TYPE', 'CNT_FAM_MEMBERS', 'EXT_SOURCE_1', 'EXT_SOURCE_2',
       'EXT_SOURCE_3', 'APARTMENTS_AVG', 'BASEMENTAREA_AVG',
       'YEARS_BEGINEXPLUATATION_AVG', 'YEARS_BUILD_AVG', 'COMMONAREA_AVG',
       'ELEVATORS_AVG', 'ENTRANCES_AVG', 'FLOORSMAX_AVG', 'FLOORSMIN_AVG',
       'LANDAREA_AVG', 'LIVINGAPARTMENTS_AVG', 'LIVINGAREA_AVG',
       'NONLIVINGAPARTMENTS_AVG', 'NONLIVINGAREA_AVG', 'APARTMENTS_MODE',
       'BASEMENTAREA_MODE', 'YEARS_BEGINEXPLUATATION_MODE', 'YEARS_BUILD_MODE',
       'COMMONAREA_MODE', 'ELEVATORS_MODE', 'ENTRANCES_MODE', 'FLOORSMAX_MODE',
       'FLOORSMIN_MODE', 'LANDAREA_MODE', 'LIVINGAPARTMENTS_MODE',
       'LIVINGAREA_MODE', 'NONLIVINGAPARTMENTS_MODE', 'NONLIVINGAREA_MODE',
       'APARTMENTS_MEDI', 'BASEMENTAREA_MEDI', 'YEARS_BEGINEXPLUATATION_MEDI',
       'YEARS_BUILD_MEDI', 'COMMONAREA_MEDI', 'ELEVATORS_MEDI',
       'ENTRANCES_MEDI', 'FLOORSMAX_MEDI', 'FLOORSMI

In [16]:
#calculating percentage of null value
null_value_percent = (current_application_data[null_columns].isnull().sum()*100/len(current_application_data)).sort_values()

In [17]:
print(null_value_percent)

DAYS_LAST_PHONE_CHANGE       0.000325
CNT_FAM_MEMBERS              0.000650
AMT_ANNUITY                  0.003902
AMT_GOODS_PRICE              0.090403
EXT_SOURCE_2                 0.214626
                              ...    
NONLIVINGAPARTMENTS_MODE    69.432963
NONLIVINGAPARTMENTS_AVG     69.432963
COMMONAREA_MODE             69.872297
COMMONAREA_AVG              69.872297
COMMONAREA_MEDI             69.872297
Length: 67, dtype: float64


In [19]:
#Finding the columns having more than 45% null values
missing_column_45= null_value_percent[null_value_percent>45].sort_values(ascending=False)

In [20]:
missing_column_45

COMMONAREA_MEDI                 69.872297
COMMONAREA_AVG                  69.872297
COMMONAREA_MODE                 69.872297
NONLIVINGAPARTMENTS_AVG         69.432963
NONLIVINGAPARTMENTS_MODE        69.432963
NONLIVINGAPARTMENTS_MEDI        69.432963
FONDKAPREMONT_MODE              68.386172
LIVINGAPARTMENTS_AVG            68.354953
LIVINGAPARTMENTS_MEDI           68.354953
LIVINGAPARTMENTS_MODE           68.354953
FLOORSMIN_MODE                  67.848630
FLOORSMIN_MEDI                  67.848630
FLOORSMIN_AVG                   67.848630
YEARS_BUILD_AVG                 66.497784
YEARS_BUILD_MEDI                66.497784
YEARS_BUILD_MODE                66.497784
OWN_CAR_AGE                     65.990810
LANDAREA_AVG                    59.376738
LANDAREA_MEDI                   59.376738
LANDAREA_MODE                   59.376738
BASEMENTAREA_MODE               58.515956
BASEMENTAREA_AVG                58.515956
BASEMENTAREA_MEDI               58.515956
EXT_SOURCE_1                    56

Inference : There are total 49 cloumns containing more than 45% null values. These columns can be removed as it it not feasible to impute missing values for these cloumns.

In [18]:
missing_column_45

Index(['YEARS_BUILD_MODE', 'YEARS_BUILD_MEDI', 'YEARS_BEGINEXPLUATATION_MODE',
       'YEARS_BEGINEXPLUATATION_MEDI', 'WALLSMATERIAL_MODE', 'TOTALAREA_MODE',
       'OBS_60_CNT_SOCIAL_CIRCLE', 'OBS_30_CNT_SOCIAL_CIRCLE',
       'NONLIVINGAREA_MODE', 'NONLIVINGAREA_MEDI', 'NONLIVINGAREA_AVG',
       'NONLIVINGAPARTMENTS_MODE', 'NONLIVINGAPARTMENTS_MEDI',
       'NONLIVINGAPARTMENTS_AVG', 'LIVINGAREA_MODE', 'LIVINGAREA_MEDI',
       'LIVINGAREA_AVG', 'LIVINGAPARTMENTS_MODE', 'LIVINGAPARTMENTS_MEDI',
       'LIVINGAPARTMENTS_AVG', 'LANDAREA_MODE', 'LANDAREA_MEDI',
       'LANDAREA_AVG', 'HOUSETYPE_MODE', 'FONDKAPREMONT_MODE',
       'FLOORSMIN_MODE', 'FLOORSMIN_MEDI', 'FLOORSMAX_MODE', 'FLOORSMAX_MEDI',
       'ENTRANCES_MODE', 'ENTRANCES_MEDI', 'EMERGENCYSTATE_MODE',
       'ELEVATORS_MODE', 'ELEVATORS_MEDI', 'DEF_60_CNT_SOCIAL_CIRCLE',
       'DEF_30_CNT_SOCIAL_CIRCLE', 'DAYS_LAST_PHONE_CHANGE', 'COMMONAREA_MODE',
       'COMMONAREA_MEDI', 'BASEMENTAREA_MODE', 'BASEMENTAREA_MEDI',
     

In [25]:
#shape before removing the columns
emptycol=current_application_data.isnull().sum()
emptycol=emptycol[emptycol.values>(0.3*len(emptycol))]
print(len(emptycol))

64


In [ ]:
# Removing the column containg more than 45% of null value

In [ ]:
#shape after removing the columns

Now we are left with 73 columns. We'll analyse the each column containing the percentange of null value less than 15%.

In [ ]:
# Finding the columns containing missing values less than 15%

In [ ]:
#Analyse AMT_REQ_CREDIT_BUREAU_WEEK column in the dataframe for missing values

In [14]:
#null value in AMT_REQ_CREDIT_BUREAU_WEEK

### Inference : 

AMT_REQ_CREDIT_BUREAU_WEEK column has 13.50% null values. These null values can be imputed with the median value i.e 0

In [ ]:
#Fill Null values in AMT_REQ_CREDIT_BUREAU_WEEK with 0

In [ ]:
#null value in AMT_REQ_CREDIT_BUREAU_WEEK

In [15]:
#null value before removing for AMT_REQ_CREDIT_BUREAU_MON

In [ ]:
#Analyse AMT_REQ_CREDIT_BUREAU_MON column in the dataframe for missing values

Inference :AMT_REQ_CREDIT_BUREAU_MON column has 13.50% null values. These null values can be imputed with the median value i.e 0


In [ ]:
#null value after removing for AMT_REQ_CREDIT_BUREAU_MON

In [ ]:
#null value before removing for AMT_REQ_CREDIT_BUREAU_DAY

In [ ]:
#Analyse AMT_REQ_CREDIT_BUREAU_DAY column in the dataframe for missing values

Inference :AMT_REQ_CREDIT_BUREAU_DAY column has 13.50% null values. These null values can be imputed with the median value i.e 0

In [ ]:
#Fill Null values in AMT_REQ_CREDIT_BUREAU_DAY with 0

In [ ]:
#null value after removing for AMT_REQ_CREDIT_BUREAU_DAY

In [ ]:
#null value before removing for AMT_REQ_CREDIT_BUREAU_HOUR

In [ ]:
#Analyse AMT_REQ_CREDIT_BUREAU_HOUR column in the dataframe for missing values


Inference : AMT_REQ_CREDIT_BUREAU_HOUR column has 13.50% null values. These null values can be imputed with the median value i.e 0 as median and 50% quantile is nearly in the same range.

In [ ]:
#Fill Null values in AMT_REQ_CREDIT_BUREAU_HOUR with 0

In [ ]:
#null value after removing for AMT_REQ_CREDIT_BUREAU_HOUR

In [ ]:
#null value before removing for AMT_REQ_CREDIT_BUREAU_QRT

In [ ]:
#Analyse AMT_REQ_CREDIT_BUREAU_QRT column in the dataframe for missing values

Inference : AMT_REQ_CREDIT_BUREAU_QRT column has 13.50% null values. These null values can be imputed with the median value i.e 0 as median and 50% quantile is nearly in the same range

In [ ]:
#Fill Null values in AMT_REQ_CREDIT_BUREAU_QRT with 0

In [ ]:
#null value after removing for AMT_REQ_CREDIT_BUREAU_QRT

In [ ]:
#null value before removing for AMT_REQ_CREDIT_BUREAU_YEAR

In [ ]:
#Analyse AMT_REQ_CREDIT_BUREAU_YEAR column in the dataframe for missing values

Inference : AMT_REQ_CREDIT_BUREAU_YEAR column has 13.50% null values. These null values can be imputed with the median value i.e 1 as median and 50% quantile is nearly in the same range.

In [ ]:
#Fill Null values in AMT_REQ_CREDIT_BUREAU_DAY with 1

In [ ]:
#null value before after for AMT_REQ_CREDIT_BUREAU_YEAR

#### Checking Columns containing more than 15% null value

In [ ]:
#Finding columns having null values more than 15%

#### We can fill the null value for OCCUPATION_TYPE based on Income

In [ ]:
#Finding number of null values in OCCUPATION_TYPE column

##### Inference : We are filling null value as per income . If the income is more than 1 Lakh then it is Managers else Laborers

In [ ]:
#filling null value based on income

#### Inference : We can see the median value as 24903 so we are filling the null value.

In [ ]:
#null value after filling

#### Inference : We are filling the null value for good price as median as both median and 50% are nearly same.

In [ ]:
# Fnd the number of null values in NAME_TYPE_SUITE column

##### Inference : We have more numbers of Unaccompanied so we are filling the null values with it.

##### Inference : We can see the mean and 50% (median) is almost same [2] . We are filling null values with 2

In [ ]:
# Find the number of null values in CNT_FAM_MEMBERS column

In [ ]:
# Find the number of null values in DAYS_LAST_PHONE_CHANGE column

### Removing Unwanted Columns

In [ ]:
#Listing columns which are not required for the dateframe study

In [ ]:

#Removing unwanted columns from the current_application_data dataframe

In [ ]:

#Checking shape of current_application_data dataframe after removing unwanted columns

Insights : 1. Removed columns having null value percentage more than 45% 2. Imputed the null values accordingly for columns containing less percentage of null values. 3. Removed unwanted data from data set.

### Checking the XNA value from table and fill it as per condition

##### Inference : Female is having the majority and only 4 rows are having NA values, we can update those columns with Gender 'F' as there will be no impact on the dataset.

In [ ]:
# Replacing XNA with F

In [ ]:
#Checking for organization type

Inference : XNA having 18% and we cant impute as per the condition.So we are removing it

### Converting Negative Values to Positive using absolute values

##### Inference : We can see there are columns having negative values so changing it to absolute values.

### Outliers in current_application_data dataframe

In [ ]:
#Analizing the outlier for AMT_CREDIT

##### Inference : As per the plot there are applicants having credit loan amount more than 3 lakhs.

In [ ]:
#Analizing the outlier for AMT_ANNUITY

##### Inference : Most of the applicants having annuity amount from 15000 to 40000 but very few are having annuity 25 Lakhs

In [ ]:
#Analizing the outlier for AMT_INCOME_TOTAL

##### Inference : Most of the applicant having the income range of 1 lakhs to 3 lakhs but there is only one applicant having income around 11-12Cr so we can consider it as outlier.

In [ ]:

#Analizing the outlier for AMT_GOOD_PRICE

##### Inference : Most of the applicants are having goods price between 2-7 Lakhs but some applicants having goods price of 40 Lakhs so it is considered as outliers.

##### Inference : Majority of the applicants are having 2-3 Members but in some applicants there are having more than 5 Members and only 1 applicant is having 20 Members . So we are considering as outliers.

### Grouping the column based on income, credit, annuity and goods cost

In [ ]:
#Grouping the income

In [ ]:
#checking the values of income group

In [ ]:
# grouping the credit

In [ ]:
#checking the values of current group

In [ ]:
#grouping the annuity

In [ ]:
#checking the values of annuity group

In [ ]:
# grouping the good price

### Target Imbalance

##### Inference : 91.3% of applicants are not facing difficulties with payment.

In [ ]:
#Calculating the impurity balance for targets

### Plotting the data based on target (0 and 1)

### Univariant plot for target 0

Insights : Female applicates count is more than male applicants count in income range 25k-125k whereas Male applicants count is more than female applicates count in income range 5 Lakhs and above

We can see the similar count for Male and Female applicatants in income range 2.5L-2.7L

##### Insights : Female applicants count is more than male applicants for all credit amount ranges.

Highest Credit Range : 10 Lakhs and Above

##### Insights : Most of the male are having greated annuity than female for the range of 1.5 L to 2.25L . female applicants are having annuity amount more in the range of 0-1.5L and 2.25L-2.5L than male

Insights : Female Applicants are more than male applicants in all goods price amount range.

Highest Credit Range : 2 lakhs to 3 Lakhs

##### Insights : Most of the people prefer for Cash Loans

Insights : Female applicants has more Cash loans type contract product than males.

##### Insights : More Female Applicants are in Working.Commercial Associates,State Servants,Pensioners and Maternity Leave whereas More male applicants are in Students and Businessman.

Insights : 1. Females applicant count is highest at Manager level. 2. Male labour applicant count is highest amoung all male applicants.

Insights : More Female applicats than that of males in all name_type_suite.

###### Insights : 1. More loan applicants are from organization type business entity type 3 and self-employed. 2. Very less number of the loan applicants from industry type 8 and trade type 5

Insights : 1. For target 0 most of the loan applicants are having occupation type Laborers. 2. For target 0 very few of the loan applicants are having occupation type IT staff

### Univariant plot for target 1

##### Insights : There are more female applicant counts in amount income range 75k-100k than that of male applicants for target 1 2. There are equal male applicant and female counts in amount income range 225k-250k for target 1